velox/functions/sparksql/String.h

```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
   }
 };
 
+/// translate(string, matchStr, replaceStr) -> varchar
+///
+///   Returns a new translated string. It translates the character in ``string``
+///   by a character in ``replaceStr``. The character in ``replaceStr`` is
+///   corresponding to the character in ``matchStr``. The translation will
+///   happen when any character in ``string`` matching with the character in the
+///   ``matchStr``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  using MapType = std::unordered_map<StringView, StringView>;
+  MapType dict_;
+  bool isConstantTranslateDict = false;
+
+  FOLLY_ALWAYS_INLINE MapType buildDict(
+      const arg_type<Varchar>& matchStr,
+      const arg_type<Varchar>& replaceStr) {
+    MapType dict;
+    int i = 0, j = 0;
+    while (i < matchStr.size()) {
+      StringView replacePiece;
+      if (j >= replaceStr.size()) {
+        replacePiece = StringView("");
+      } else {
+        int replaceCharCount = utf8proc_char_length(replaceStr.data() + j);
+        replacePiece = StringView(replaceStr.data() + j, replaceCharCount);
+        j = j + replaceCharCount;
+      }
+      int matchCharCount = utf8proc_char_length(matchStr.data() + i);
+      StringView matchPiece = StringView(matchStr.data() + i, matchCharCount);
+      dict.try_emplace(matchPiece, replacePiece);
```

Comment:
Should we check if there are duplicates and throw? What does Spark do in this case? Let's document.

Reply:
Spark only considers the firstly occurred character in match string. So we dismiss the match character and its corresponding replace character if the match character already exists in the built dictionary.